File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
       if(num.size() == k) return "0";
       int size = k;
       stack<int>st; string ans; int i=0;
       while(num[i] != '\0'){
  
            while(!st.empty() && st.top() > (int)num[i] && size >0){
              st.pop(); size--;
            }
            if(size == 0) break;
            st.push((int)num[i]);
      
         i++;
       }
       while(i<num.size()){
        st.push((int)num[i]);
        i++;
       }

       if(size > 0){
        while(size--){
            st.pop();
        }
       }

       while(!st.empty()){
        ans += (char)st.top();
        st.pop();
       }

       reverse(ans.begin(),ans.end());
       i=0;
       if(ans[i] == '0'){
       while(ans[i] == '0'){
        i++;
       }
       ans = ans.substr(i);
       }
       if(ans.size() == 0) return "0";
       return ans;


    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        if(k == num.length() || k > num.length()) return "0";
        Stack<Character>st = new Stack<>();

        for(int i=0;i<num.length();i++){
          char c = num.charAt(i);
          if(st.size() == 0)st.add(c);
          else{
            while(st.size() > 0 && k > 0 && st.peek() > c){
                st.pop();
                k--;
            }
            st.add(c);
          }
        }
        while(k>0){
            st.pop();
            k--;
        }

        StringBuilder str = new StringBuilder();
        while(st.size() > 0){
            str.append(st.pop());
        }
        str = str.reverse();
        int i=0;
        boolean zero = false;
        while(i<str.length() && str.charAt(i) == '0'){
         i++;
         zero = true;
        }
       if(i == str.length()) return "0";
        return str.toString().substring(i,str.length());
    }
}